### PR TITLE
Hide types which are now public in corefx

### PR DESCRIPTION
**Pull request by rahku**
When this change hits corefx it requires change in src\System.Threading\src\System.Threading.csproj to use Environment type in System.Runtime.Extensions instead of corelib.

### CODE REVIEW

**Comment by rahku** [comment]
PTAL @jkotas

File: src/mscorlib/src/System/Collections/CollectionBase.cs [hunk 1]
```diff
@@ -13,7 +13,7 @@ namespace System.Collections
 {
     // Useful base class for typed read/write collections where items derive from object
     [Serializable]
-    public abstract class CollectionBase : IList
+    internal abstract class CollectionBase : IList
```
Comment (jkotas) [inline src/mscorlib/src/System/Collections/CollectionBase.cs]: Delete it instead - it is not used in CoreLib.

File: src/mscorlib/src/System/Collections/ObjectModel/KeyedCollection.cs [hunk 2]
```diff
@@ -14,7 +14,7 @@ namespace System.Collections.ObjectModel
     [Serializable]
     [DebuggerTypeProxy(typeof(Mscorlib_KeyedCollectionDebugView<,>))]
     [DebuggerDisplay("Count = {Count}")]
-    public abstract class KeyedCollection<TKey, TItem> : Collection<TItem>
+    internal abstract class KeyedCollection<TKey, TItem> : Collection<TItem>
```
Comment (jkotas) [inline src/mscorlib/src/System/Collections/ObjectModel/KeyedCollection.cs]: Delete

File: src/mscorlib/src/System/Collections/StructuralComparisons.cs [hunk 3]
```diff
@@ -8,7 +8,7 @@
 
 namespace System.Collections
 {
-    public static class StructuralComparisons
+    internal static class StructuralComparisons
```
Comment (jkotas) [inline src/mscorlib/src/System/Collections/StructuralComparisons.cs]: Delete

File: src/mscorlib/src/System/IServiceObjectProvider.cs [hunk 4]
```diff
@@ -8,7 +8,7 @@
 
 namespace System
 {
-    public interface IServiceProvider
+    internal interface IServiceProvider
```
Comment (jkotas) [inline src/mscorlib/src/System/IServiceObjectProvider.cs]: Delete

File: src/mscorlib/src/System/Reflection/RuntimeReflectionExtensions.cs [hunk 5]
```diff
@@ -7,7 +7,7 @@
 
 namespace System.Reflection
 {
-    public static class RuntimeReflectionExtensions
+    internal static class RuntimeReflectionExtensions
```
Comment (jkotas) [inline src/mscorlib/src/System/Reflection/RuntimeReflectionExtensions.cs]: Delete

File: src/mscorlib/src/System/Runtime/CompilerServices/CompilerMarshalOverride.cs [hunk 6]
```diff
@@ -17,7 +17,7 @@ namespace System.Runtime.CompilerServices
     //
     // Indicates that the modified instance differs from its true type for
     // correct marshalling.
-    public static class CompilerMarshalOverride
+    internal static class CompilerMarshalOverride
```
Comment (jkotas) [inline src/mscorlib/src/System/Runtime/CompilerServices/CompilerMarshalOverride.cs]: Delete

File: src/mscorlib/src/System/Runtime/CompilerServices/SpecialNameAttribute.cs [hunk 7]
```diff
@@ -17,7 +17,7 @@ namespace System.Runtime.CompilerServices
                     AttributeTargets.Struct)]
 
 
-    public sealed class SpecialNameAttribute : Attribute
+    internal sealed class SpecialNameAttribute : Attribute
```
Comment (jkotas) [inline src/mscorlib/src/System/Runtime/CompilerServices/SpecialNameAttribute.cs]: Delete

File: src/mscorlib/src/System/Security/SecurityState.cs [hunk 8]
```diff
@@ -7,7 +7,7 @@
 
 namespace System.Security
 {
-    public abstract class SecurityState
+    internal abstract class SecurityState
```
Comment (jkotas) [inline src/mscorlib/src/System/Security/SecurityState.cs]: Delete

File: src/mscorlib/src/System/Threading/CountdownEvent.cs [hunk 9]
```diff
@@ -29,7 +29,7 @@ namespace System.Threading
     /// accessing the event.
     /// </remarks>
     [DebuggerDisplay("Initial Count={InitialCount}, Current Count={CurrentCount}")]
-    public class CountdownEvent : IDisposable
+    internal class CountdownEvent : IDisposable
```
Comment (jkotas) [inline src/mscorlib/src/System/Threading/CountdownEvent.cs]: Delete

**Comment by jkotas** [comment]
I think you will need to wait for https://github.com/dotnet/coreclr/pull/10074 to fix the test failures

**Comment by jkotas** [comment]
>  I need to merge this PR in order to build System.Threading.Thread correctly.

I do not think you need to. You have fixed corefx to the right thing already - you just need to pick up a new corefx build.

File: src/mscorlib/src/System/Diagnostics/Stacktrace.cs [hunk 10]
```diff
@@ -388,7 +388,7 @@ public StackTrace(StackFrame frame)
         // Constructs a stack trace for the given thread
         //
         [Obsolete("This constructor has been deprecated.  Please use a constructor that does not require a Thread parameter.  http://go.microsoft.com/fwlink/?linkid=14202")]
-        public StackTrace(Thread targetThread, bool needFileInfo)
+        internal StackTrace(Thread targetThread, bool needFileInfo)
```
Comment (jkotas) [inline src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: Can this be deleted?

**Comment by rahku** [comment]
Not System.Threading.Thread but System.Threading. https://github.com/dotnet/corefx/blob/master/src/System.Threading/src/System.Threading.csproj#L43

System.Threading.csproj uses Environment from corelib instead of runtime.extensions. This has not been fixed it. I get errors for duplicate types in corelib & runtime.extensions if I fix it now. I know there is some way to fix this in corefx without this PR. But can do after this change hits corefx as well.

**Comment by jkotas** [comment]
ok

**Comment by rahku** [comment]
@dotnet-bot test Windows_NT x64 Debug Build and Test

**Comment by rahku** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test
@dotnet-bot test OSX x64 Checked Build and Test

**Comment by rahku** [comment]
@jkotas does this look fine

**Comment by rahku** [comment]
@dotnet-bot test OSX x64 Checked Build and Test

**Pull request merged**